File: soc-server/internal/models/case.go
// ==============================================================================
// Package models - GORM Model: Case
// File: internal/models/case.go
// Mô tả: Bảng `cases` quản lý các vụ việc bảo mật (Security Incidents).
//         Thay thế chức năng TheHive trong kiến trúc SOC truyền thống.
//         Mỗi Case có thể được tạo từ 1 hoặc nhiều Alert liên quan.
// ==============================================================================

package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// CaseStatus - Enum trạng thái vòng đời của Case
type CaseStatus string

const (
	CaseStatusNew        CaseStatus = "New"        // Mới tạo, chưa phân tích
	CaseStatusInProgress CaseStatus = "InProgress" // Đang được analyst điều tra
	CaseStatusClosed     CaseStatus = "Closed"     // Đã xử lý và đóng
	CaseStatusRejected   CaseStatus = "Rejected"   // Từ chối (False Positive)
)

// Case - GORM model cho bảng `cases`
// Đại diện cho một vụ việc bảo mật cần điều tra và xử lý
type Case struct {
	// ID - Khóa chính UUID
	ID string `gorm:"size:36;primaryKey" json:"id"`

	// AlertID - FK liên kết tới Alert gốc đã sinh ra Case này
	AlertID string `gorm:"size:36;index" json:"alert_id"`

	// Title - Tiêu đề vụ việc (ví dụ: "Ransomware Activity Detected on WS-01")
	Title string `gorm:"type:varchar(500);not null" json:"title"`

	// Description - Mô tả chi tiết vụ việc, hỗ trợ Markdown
	// Analyst có thể ghi chú quá trình điều tra vào đây
	Description string `gorm:"type:text" json:"description"`

	// SeverityNum - Mức độ nghiêm trọng dạng số (1=Critical, 2=High, 3=Medium, 4=Low)
	// Dùng số để dễ sort và filter trên Dashboard
	SeverityNum int `gorm:"type:integer;default:3;index" json:"severity_num"`

	// Status - Trạng thái vòng đời Case
	Status CaseStatus `gorm:"type:varchar(30);default:'New';index" json:"status"`

	// AssignedTo - Username/email của analyst được giao phụ trách Case
	AssignedTo string `gorm:"type:varchar(255)" json:"assigned_to"`

	// Tags - Nhãn phân loại dạng JSONB array (ví dụ: ["ransomware", "apt", "internal"])
	// Cho phép gắn nhiều tags để filter và search linh hoạt
	Tags string `gorm:"type:json" json:"tags"`

	// SOARStatus - Trạng thái xử lý SOAR (pending, ai_analyzing, human_review, completed)
	SOARStatus string `gorm:"type:varchar(50);default:'pending'" json:"soar_status"`

	// AIReason - Lý do/kết luận từ AI analysis (nếu có)
	AIReason string `gorm:"type:text" json:"ai_reason"`

	// HumanApprovedBy - Người đã phê duyệt qua Telegram HITL (nếu có)
	HumanApprovedBy string `gorm:"type:varchar(255)" json:"human_approved_by"`

	// Confidence - Độ tin cậy của kết quả phân tích (0.0 - 1.0)
	Confidence float64 `gorm:"type:decimal(5,4);default:0" json:"confidence"`

	// Timestamps
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`

	// Relations
	Alert *Alert `gorm:"foreignKey:AlertID" json:"alert,omitempty"`
}

// TableName - Chỉ định tên bảng
func (Case) TableName() string {
	return "cases"
}

// BeforeCreate - Tự động sinh UUID
func (c *Case) BeforeCreate(tx *gorm.DB) error {
	if c.ID == "" {
		c.ID = uuid.New().String()
	}
	if c.Status == "" {
		c.Status = CaseStatusNew
	}
	return nil
}

File: soc-server-python/app/models/case.py
# ==============================================================================
# app/models/case.py - SQLAlchemy Model: Case
# Tương đương: internal/models/case.go
# ==============================================================================

import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from sqlalchemy import String, DateTime, Text, Integer, Float, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class CaseStatus(str, Enum):
    NEW         = "New"
    IN_PROGRESS = "InProgress"
    CLOSED      = "Closed"
    REJECTED    = "Rejected"


class Case(Base):
    __tablename__ = "cases"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))

    # FK → Alert
    alert_id: Mapped[Optional[str]] = mapped_column(String(36), ForeignKey("alerts.id"), index=True, nullable=True)

    # Tiêu đề vụ việc
    title: Mapped[str] = mapped_column(String(500), nullable=False)

    # Mô tả (hỗ trợ Markdown)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Mức nghiêm trọng: 1=Critical, 2=High, 3=Medium, 4=Low
    severity_num: Mapped[int] = mapped_column(Integer, default=3, index=True)

    # Trạng thái vòng đời
    status: Mapped[str] = mapped_column(String(30), default="New", index=True)

    # Analyst được giao
    assigned_to: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)

    # Tags dạng JSON array
    tags: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # SOAR processing
    soar_status: Mapped[str] = mapped_column(String(50), default="pending")
    ai_reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    human_approved_by: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    confidence: Mapped[float] = mapped_column(Float, default=0.0)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    deleted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True, index=True)

    # Relations
    alert: Mapped[Optional["Alert"]] = relationship("Alert", foreign_keys=[alert_id], lazy="select")  # type: ignore
